File: WorldFeed/Server/src/WorldFeed.Common/DTO/UploadProcessingInfo.cs
﻿namespace WorldFeed.Common.DTO
{
    using Newtonsoft.Json;

    using WorldFeed.Common.Public.Models.Enums;
    using WorldFeed.Common.Public.Models.Interfaces.DTO;

    public class UploadProcessingInfo : IUploadProcessingInfo
    {
        public UploadProcessingInfo()
        {
            CheckAfterInSeconds = 0;
            ProgressPercentage = 0;
        }

        [JsonProperty("state")]

        public string State { get; set; }

        [JsonIgnore]
        public ProcessingState ProcessingState
        {
            get
            {
                switch (State)
                {
                    case "pending":
                        return ProcessingState.Pending;
                    case "in_progress":
                        return ProcessingState.InProgress;
                    case "succeeded":
                        return ProcessingState.Succeeded;
                    case "failed":
                        return ProcessingState.Failed;
                    default:
                        return ProcessingState.Undefined;
                }
            }
            set
            {
                switch (value)
                {
                    case ProcessingState.Pending:
                        State = "pending";
                        break;
                    case ProcessingState.InProgress:
                        State = "in_progress";
                        break;
                    case ProcessingState.Succeeded:
                        State = "succeeded";
                        break;
                    case ProcessingState.Failed:
                        State = "failed";
                        break;
                    case ProcessingState.Undefined:
                        State = "undefined";
                        break;

                }
            }
        }

        [JsonProperty("check_after_secs")]
        public int CheckAfterInSeconds { get; set; }

        [JsonIgnore]
        public int CheckAfterInMilliseconds => CheckAfterInSeconds * 1000;

        [JsonProperty("progress_percent")]
        public int ProgressPercentage { get; set; }

        [JsonProperty("error")]
        public IUploadProcessingError Error { get; set; }
    }
}

File: WorldFeed/Client/src/app/core/Core/DTO/UploadProcessingInfo.ts
import {ProcessingState} from "../../Public/Models/Enum/ProcessingState";
import {IUploadProcessingInfo} from "../../Public/Models/Interfaces/DTO/IUploadProcessingInfo";
import {IUploadProcessingError} from "../../Public/Models/Interfaces/DTO/IUploadProcessingError";

export class UploadProcessingInfo implements IUploadProcessingInfo {
  constructor() {
    this.checkAfterInSeconds = 0;
    this.progressPercentage = 0;
  }

  // [JsonProperty("state")]
  public state: string;

  get processingState(): ProcessingState {
    switch (this.state) {
      case "pending":
        return ProcessingState.Pending;
      case "in_progress":
        return ProcessingState.InProgress;
      case "succeeded":
        return ProcessingState.Succeeded;
      case "failed":
        return ProcessingState.Failed;
      default:
        return ProcessingState.Undefined;
    }
  }

  set processingState(value: ProcessingState) {
    switch (value) {
      case ProcessingState.Pending:
        this.state = "pending";
        break;
      case ProcessingState.InProgress:
        this.state = "in_progress";
        break;
      case ProcessingState.Succeeded:
        this.state = "succeeded";
        break;
      case ProcessingState.Failed:
        this.state = "failed";
        break;
      case ProcessingState.Undefined:
        this.state = "undefined";
        break;
    }
  }

  // [JsonProperty("check_after_secs")]
  public checkAfterInSeconds: number;

  // [JsonIgnore]
  get checkAfterInMilliseconds(): number {
    return this.checkAfterInSeconds * 1000;
  }

  // [JsonProperty("progress_percent")]
  public progressPercentage: number;

  // [JsonProperty("error")]
  public error: IUploadProcessingError;
}
